src/Shared/FileMatcher.cs

```diff
@@ -142,16 +147,16 @@ private static string[] GetAccessibleFileSystemEntries(FileSystemEntity entityTy
         /// </summary>
         /// <param name="path"></param>
         /// <param name="pattern"></param>
-        /// <returns>Array of matching file system entries (can be empty).</returns>
-        private static string[] GetAccessibleFilesAndDirectories(string path, string pattern)
+        /// <returns>An enumerable of matching file system entries (can be empty).</returns>
+        private static IEnumerable<string> GetAccessibleFilesAndDirectories(string path, string pattern)
         {
-            string[] entries = null;
-
             if (Directory.Exists(path))
             {
                 try
                 {
-                    entries = Directory.GetFileSystemEntries(path, pattern);
+                    return ShouldEnforceMatching(pattern)
+                        ? Directory.EnumerateFileSystemEntries(path, pattern).Where(o => IsMatch(Path.GetFileName(o), pattern))
```

Comment:
You need to think about exception semantics, no longer are exceptions isolated here - they can be thrown when the "enumerable" is iterated.

Reply:
Ouch, I completely missed that. I will wrap the enumerator into a safe enumerator that will omit those two exceptions, in order to preserve the same behavior, prior this change.

Reply:
> no longer are exceptions isolated here

Sadly, this is not actually true, when `EnumerateFileSystemEntries`/`EnumerateFiles`/`EnumerateDirectories` methods are called, they can also throw because the file enumerator is initialized by finding the first file/directory and if an error occurs, the exception is [immediately thrown](https://github.com/dotnet/coreclr/blob/da8b7181dbe362fbfe64c7ef72ff3a4a7db68f24/src/mscorlib/src/System/IO/FileSystemEnumerable.cs#L217). So we have to handle exceptions when one of those three methods is called and on the MoveNext call.